stackslib/src/chainstate/stacks/boot/signers-voting.clar

```diff
@@ -0,0 +1,102 @@
+    (burn-height-to-reward-cycle burn-block-height))
+
+(define-read-only (get-last-round (reward-cycle uint))
+    (map-get? rounds reward-cycle))
+
+(define-read-only (get-vote (reward-cycle uint) (round uint) (signer principal))
+    (map-get? votes {reward-cycle: reward-cycle, round: round, signer: signer}))
+
+(define-read-only (get-signer-slots (signer-index uint) (reward-cycle uint))
+    (let ((height (reward-cycle-to-burn-height reward-cycle)))
+            (ok (at-block
+                (unwrap! (get-block-info? id-header-hash height) err-invalid-burn-block-height)
+                    (get-current-signer-slots signer-index)))))
+
+(define-read-only (get-current-signer-slots (signer-index uint))
+    (let ((details (unwrap! (unwrap-panic (contract-call? .signers stackerdb-get-signer-by-index signer-index)) err-invalid-signer-index)))
+        (asserts! (is-eq (get signer details) tx-sender) err-signer-index-mismatch)
+        (ok (get num-slots details))))
+
+;; aggregate public key must be unique and can be used only in a single cycle-round pair
+(define-read-only (is-valid-aggregated-public-key (key (buff 33)) (dkg-id {reward-cycle: uint, round: uint}))
+    (is-eq (default-to dkg-id (map-get? used-aggregate-public-keys key)) dkg-id))
+
+(define-read-only (is-in-prepare-phase (height uint))
+    (< (mod (+ (- height (get first-burnchain-block-height pox-info))
+                (get prepare-cycle-length pox-info))
+             (get reward-cycle-length pox-info)
+            )
+        (get prepare-cycle-length pox-info)))
+
+(define-private (is-in-voting-window (height uint) (reward-cycle uint))
+    (let ((last-cycle (unwrap-panic (contract-call? .signers stackerdb-get-last-set-cycle))))
+        (and (is-eq last-cycle reward-cycle)
+            (is-in-prepare-phase height))))
+
+(define-public (vote-for-aggregate-public-key (signer-index uint) (key (buff 33)) (round uint))
+    (let ((reward-cycle (+ u1 (burn-height-to-reward-cycle burn-block-height)))
+            (tally-key {reward-cycle: reward-cycle, round: round, aggregate-public-key: key})
+            ;; one slot, one vote
+            (num-slots (try! (get-current-signer-slots signer-index)))
+            (new-total (+ num-slots (default-to u0 (map-get? tally tally-key)))))
+        (asserts! (is-in-voting-window burn-block-height reward-cycle) err-out-of-voting-window)
+        (asserts! (>= round (default-to u0 (map-get? rounds reward-cycle))) err-old-round)
+        (asserts! (is-eq (len key) u33) err-invalid-aggregate-public-key)
+        (asserts! (is-valid-aggregated-public-key key {reward-cycle: reward-cycle, round: round}) err-duplicate-aggregate-public-key)
+        (asserts! (map-insert votes {reward-cycle: reward-cycle, round: round, signer: tx-sender} {aggregate-public-key: key, reward-slots: num-slots}) err-duplicate-vote)
+        (map-set tally tally-key new-total)
+        (map-set used-aggregate-public-keys key {reward-cycle: reward-cycle, round: round})
```

Comment:
Shouldn't this be set only once consensus is reached / the key is set? Currently, I think no two signers would be able to vote for the same key.

Reply:
They should always vote for the same key in the same round.

Reply:
Correct. I'm saying the say this (map-set ...) is set up *every* vote adds a key to the "is-valid-aggregated-public-key" no?

So let's say Alice votes for the correct aggregate public key X, with the current setup X is automatically added as a used-aggregate-public-key. 

So when Bob comes along to also vote for the correct aggregate public key X, he **won't** be able to & will get an err-duplicate-aggregate-public-key error.

Reply:
is duplicate pub key function accepts the same key for the same cycle and round.